### PR TITLE
Add option to pass different build type in CMake build helper.

### PR DESCRIPTION
**Pull request by bobeff**
Add additional argument build_type for **CMake** build helper - build, install and test methods to be able to override **Conan** _build_type_ setting.

When using **CMake** build helper for _Release_ **Conan** setting in recipe sometimes I want to build _RelWithDebInfo_ **CMake** configuration instead of _Release_ configuration. Currently when I have _Release_ in settings and I call `cmake.build()`, **Conan** tries `cmake --build <build_dir> --config Release`. It will be good to be possible to override the default assumptions for _Debug_ when I have _Debug_ **Conan** setting and Release when I have _Release_ **Conan** setting.

Of course I always can call appropriate build command with `self.run` but having direct support for this in **CMake** build helper will be more convenient.

### CODE REVIEW

**Comment by memsharded** [comment]
I don't get the full use case. How would your full ``build()`` method look like with such CMake helper? Won't you prefer to directly add those configurations to the ``settings.yml`` build_type?

**Comment by bobeff** [comment]
I tested it with this code:

```python
cmake_build_type = self.settings.build_type
if cmake_build_type == "Release":
    cmake_build_type = "RelWithDebInfo"

cmake.build(build_type=cmake_build_type)
cmake.install(build_type=cmake_build_type)
```

If I add _RelWithDebInfo_ to `settings.yml` doesn't this means that this setting must match for packages to which my package depends and to packages which depend to my package? If so this is problem because for them version with debug info can be in _Release_ package and it is possible that they have not _RelWithDebInfo_ version.

I also thinking about adding `cmake_build_type` argument to `CMake` class constructor for overriding default `self.build_type`. Something like:

```python
def __init__(self, settings_or_conanfile, generator=None, cmake_system_name=True,
             parallel=True, build_type=None):
```
```python
self._build_type = self._settings.get_safe("build_type")
    if build_type:
        self._build_type = build_type
```
and then the usage will be simply:
```python
cmake = CMake(self, build_type="RelWithDebInfo")
cmake.configure(defs=self.__get_cmake_definitions(),
                source_dir=source_directory,
                build_dir=build_directory)
cmake.build()
cmake.install()
```

**Comment by memsharded** [comment]
Thanks for the feedback. 
So basically, you want to build all packages in Release, but then you want to force a specific package to build always with RelWithDebInfo? And are you going to deploy such package to production built with RelWithDebInfo configuration? You don't want to have a regular Release binary for that package?

If it is not hardcoded, and you want to be able to control that setting without having to modify the recipe, then you want to add those as settings in ``settings.yml``. Then, when you want to force one specific package for some reason to build with that info, you can always use the per-package settings, something like:

```bash
$ conan install . -s build_type=Release -s MyPkg:build_type=RelWithDebInfo
```
You can also specify those per-package settings in profiles.

**Comment by bobeff** [comment]
10x for the suggestion. I didn't know that  it is possible to specify per-package setting. I thought that this is available only for options and that the package settings must always match.

> You don't want to have a regular Release binary for that package?

I think that regular _Release_ binary is pretty useless, because if I had production crash in some of the third parties it will be impossible for me  to investigate it. I always build my _Release_ binaries with debug information enabled. I also think that this is **CMake** design mistake to provide _Release_ and _RelWithDebInfo_ configurations instead of _Release_ and _RelWithoutDebInfo_. :) For my own projects I remove _MinSizeRel_ and _RelWIthDebInfo_ configurations and I manually add required flags for producing debug information to _Release_ configuration. Doing so for all third parties by modding their build scripts will be highly unpractical and for this reason I prefer to build _RelWithDebInfo_ and to add it to _Release_ package.

Per-package settings solves the problem, but it is a bit cumbersome to specify such per package settings for all packages which are built using  **CMake**. For now I will use something like:

```python
cmake_build_type = self.settings.build_type
if cmake_build_type == "Release":
    cmake_build_type = "RelWithDebInfo"

build_command = "cmake --build %s --config %s" % (build_directory, cmake_build_type)
install_command = "%s --target install" % build_command

self.run(build_command)
self.run(install_command)
```

But I still think that it will be good to be possible to do this when using **CMake** build helper.

**Comment by memsharded** [comment]
> Release and RelWithoutDebInfo. 

Good point :) Though I think there are other concerns: the Debug info is leaking critical information that helps to reverse-engineer your code, which is unacceptable for many vendors releasing products to end-users.

> Per-package settings solve the problem, but it is a bit cumbersome to specify such per package settings for all packages

But if you add the settings, and use ``RelWithDebInfo`` for all your packages, then you will be good without having to customize your recipes. You still want to "label" your packages with "Release" setting instead of the "RelWithDebInfo"? Otherwise, you don't need to specify per-package settings, you can build all of them with the "RelWithDebInfo" setting.

> But I still think that it will be good to be possible to do this when using CMake build helper.

My main concern here is consistency and support, we are already overwhelmed... :) 
If we provide an easy and documented way to directly change the build_type in the CMake helper, users will use it to manually hardcode "Debug" or "Release" which won't match the ``setting.build_type``, and then later we will have a nightmare of support and remote debugging to know why some packages fail to link or even worse, at runtime, when consumed from other packages...

**Comment by bobeff** [comment]
> the Debug info is leaking critical information that helps to reverse-engineer your code

Only if I provide `pdb` files with the debug info to the end users on **Windows** or if I don't `strip`  it from final executable on  **Linux**. It is common practice to build with debug info for release, but to hold `pdb` files and unstripped executables in house for investigating production crashes when core dump is uploaded. For example default **Visual Studio** _Release_ configuration uses `/Zi` flag for producing debug info in `pdb` file.

>  You still want to "label" your packages with "Release" setting instead of the "RelWithDebInfo"? Otherwise, you don't need to specify per-package settings, you can build all of them with the "RelWithDebInfo" setting.

Yes but my consumer project is with _Release_ configuration with manually added flags for producing debug information. I admit that the reason for this is little stupid. :) It is because typing **CMAKE_CXX_FLAGS_RELWITHDEBINFO**,  **CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO** and so on is a bit more cumbersome and ugly compared to **CMAKE_CXX_FLAGS_RELEASE** ans so on ... This is cumbersome in **Conan** recipe too when typing _RelWithDebInfo_ instead of just _Release_ everywhere I have to check the `build_type` setting.

> to manually hardcode "Debug" or "Release" which won't match the setting.build_type

I don't think that exact matching some build system default configuration with package settings is completely fine because I already provided example that default _Release_ configuration of new **Visual Studio** project is doing something different providing `/Zi` flag.  Some other build system can have it's own assumptions what _Release_ means which match neither **CMake** nor **Visual Studio**. At the end the creator of the packages still be obligated manually to provide some consistency using his knowledge what _Release_ means for different build systems.

**Comment by memsharded** [comment]
I agree with most of your comments :)

> if I don't strip it from final executable on Linux.

The thing is the default, if you don't remember to strip it from the final executable, then you are leaking it. 

>  I admit that the reason for this is little stupid.

The reason is not stupid at all, I really hate typing those difficult names too!

> At the end the creator of the packages still be obligated manually to provide some consistency

Yes, totally agree. But not enabling ways to easily produce incompatibilities is to be considered here. Adding some flags or small variants for Release, doesn't generate incompatibilities, while something that will look natural for some users like:

```python
cmake = CMake(self, build_type="Debug")
cmake.configure()
cmake.build()
```
will most likely generate incompatibilities when installed with the ``Release`` setting. The normal thing is that most users won't do that, but if just 1 out of 100 users uses this feature wrongly, we will be getting many, many support requests that will take a long time to solve each.

The thing is that I agree with you, I am a big fan of the python zen "we (programmers) are all consenting adults", but the truth is that we are overwhelmed, so I tend to be more careful regarding features that could produce issues if misused. Lets see what @lasote thinks about this one. 

Thanks a lot for your feedback!

**Comment by birsoyo** [comment]
I also go to great lengths to generate debug information on release builds cmake or not. Having debug information for Release binaries is essential. I just want to mention, conan makes these *relatively easy* to handle. Thanks for this great tool.

If you add the required flags to env. variables CFLAGS and LD_FLAGS they should be picked up by cmake even on windows. So this seems to work for me (not fully tested on production code yet),
```
  with tools.environment_append({'CFLAGS':'/Zi', 'LD_FLAGS':'/DEBUG:FULL'}):
    cmake = CMake(self, generator='Ninja')
```
But note that even with cmake, getting debug info generated is not straightforward, due to both cmake itself and how CMakeLists.txt's are developed.

Examples from my cmake releated modifications for various conan packages.

libjpeg: Sometimes you have change pdb name (especially for static libs) otherwise they will be something generic like vc140.pdb.
```
  tools.replace_in_file('libjpeg/CMakeLists.txt', 
    'set_target_properties(turbojpeg-static PROPERTIES OUTPUT_NAME turbojpeg)',
    'set_target_properties(turbojpeg-static PROPERTIES OUTPUT_NAME turbojpeg)\nelse()\nset_target_properties(turbojpeg-static PROPERTIES COMPILE_PDB_NAME "turbojpeg-static" COMPILE_PDB_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}")')
[snip]
  with tools.environment_append({'CFLAGS':'/Zi', 'LD_FLAGS':'/DEBUG:FULL'}):
    cmake = CMake(self, generator='Ninja')

```
AssImp: If building with VS you can use some VS macros like OutDir, TargetName:
```
  with tools.environment_append({
    'CFLAGS':f'/Zi /Fd$(OutDir)$(TargetName).pdb /fp:fast /arch:AVX2',
    'CXXFLAGS':f'/Zi /Fd$(OutDir)$(TargetName).pdb /fp:fast /arch:AVX2',
    'LD_FLAGS':'/DEBUG:FULL'}):
    cmake.configure(source_dir=f'{self.conanfile_directory}/assimp', 
                                build_dir=f'{self.conanfile_directory}/build',
                                defs=defs)
```
And since CMake won't 'install' pdb files, you have to hunt them down in your build folder and package them correctly for conan:
```
    def package(self):
        self.copy('*', src='install', dst='', keep_path=True)
        self.copy('*.pdb', src='build/code', dst='lib', keep_path=False)
```

Even on non cmake projects:
icu:
```
  tools.replace_in_file('icu/source/tools/toolutil/Makefile.in', '$(LIBCPPFLAGS)', f'$(LIBCPPFLAGS) -Fdsicutu{suffix}.pdb')
  tools.replace_in_file('icu/source/runConfigureICU', '-Gy -MD', '-Zi -Gy -MD')
  tools.replace_in_file('icu/source/runConfigureICU', 'DEBUG_LDFLAGS=\'-DEBUG\'', 'DEBUG_LDFLAGS=\'-DEBUG\'\nRELEASE_LDFLAGS=\'-DEBUG\'')
```

boost:
```
  flags.append('debug-store=database')
  flags.append('debug-symbols=on')
```

**Comment by lasote** [comment]
I'm totally OK if we provide a ``build_type`` in the ``CMake`` constructor. I'm not afraid of people messing with the parameter. If they check the docs will see a warning, that's all what we should do.

But, some observations:

1. I would like to keep the parameter only in the constructor, not in the rest of methods.
2. Create a getter/setter for the attribute `self.build_type`. In the setter function, if the value assigned doesn't match with the settings one, print a warn message in the ouput. This way if someone want to change the attribute after creating the class can use the setter.
3. So, in the constructor one line `self.build_type = build_type or self._settings.get_safe("build_type")`
4. So we warn in the docs, we warn in the ouput, more than enough for me.

**Comment by bobeff** [comment]
Now one of the builds (on `Python 2.7 Xcode:xcode8.3`) has failed but the error message is:
> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

I think that this is not caused by my last changes, but it is due to some build machine problem. Could someone rerun the test?

**Comment by lasote** [comment]
I've relaunched it. 
Interesting that the old build_type property and the runtime were broken. Definitely, we can rename it. It looks good to me!

**Comment by lasote** [comment]
We should add a test, but the implementation is ok.

**Comment by bobeff** [comment]
I will write a test, but I'm not sure where to put it and what exactly to test. Only setting of the property or full build with different **CMake** build type? If full build test is required may be `build_test.py` is the right place?

**Comment by lasote** [comment]
It's ok. I will add it and will make you a PR to your branch

**Comment by lasote** [comment]
https://github.com/bobeff/conan/pull/1

File: conans/client/cmake.py [hunk 1]
```diff
@@ -213,18 +231,14 @@ def command_line(self):
...
-
     def _build_type_definition(self):
         if self._build_type and not self.is_multi_configuration:
             return {'CMAKE_BUILD_TYPE': self._build_type}
         return {}
 
     @property
     def runtime(self):
-        return self._defs_to_string(self._runtime_definition())
+        return _defs_to_string(self._runtime_definition())
```
Comment (memsharded) [inline conans/client/cmake.py]: +1 for catching this

**Pull request merged**